### DISCUSSION TITLE
How to search python function definitions across all installed packages

### DISCUSSION THREAD

**Discussion by mmohdnowfal** [microsoft/pylance-release#3187] [Enhancement]
I want to search for a function name across all installed packages and also in the packages defined by python.autoComplete.extraPaths setting.

Vscode search panel searches only in the current workspace. Whereas in pycharm IDE ctrl + click on a word is able to search across all packages.

Is there a way to search for function definition across all installed packages using vscode IDE?

**Comment by judej** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.

**Comment by mmohdnowfal** [comment]
Another use case where this is useful is, in case of function arguments which gets dependency injected on runtime. Different frameworks supports DI of function arguments for example Django, pylance language server will not be able to determine the type of these function variable, hence method resolutions will not work. 
So In these cases, if we can provide the full search across all the packages that will improve the experience a lot. Also vscode will be able to match what pycharm provides.

**Comment by mmohdnowfal** [comment]
It will be good if Peek --> Peek Definition on any python function/method call, should provide the definitions across all the available python libraries, this will be useful if objects type is unkown particularly the once get injected during run time

**Reply by mmohdnowfal** [reply]
pycharm provides this option, where if its not able to identify the instance type when we do goto definition on a method, then it list all the available functions/methods which maches the name

**Comment by mmohdnowfal** [comment]
This feature is mainly needed for python language, unlike c and javascript, python package dependencies can be present outside the workspace.
We should provide an option to search for any function name across all packages, vscode search will work for only workspace, whereas here we need a way to search in pip install path, and  paths pointed by `python.autoComplete.extraPaths` settings. 

@judej  Let me know your comments

**Comment by adam-kral** [comment]
So what is the workaround? I tried symlinking the library from site-packages to the workspace, but still I cannot find the symbols

**Reply by adam-kral** [reply]
Ok it works, but you should not put it here too: python.analysis.exclude

Originally I put there the whole site-packages, but it crashed pylance... So I tried the python.analysis.exclude setting.

If someone however knows how to make it work with the whole site-packages please let us know!

**Comment by rchiodo** [comment]
We've just submitted an implementation of this feature behind a new flag (coming in our next Prerelease).

`python.analysis.includeVenvInWorkspaceSymbols`. 

Setting that flag to true will also search all of your 3rd party libraries for symbols (as well as files in the current workspace). 

It uses the [python.analysis.packageIndexDepth](https://github.com/microsoft/pylance-release/blob/main/docs/settings/python_analysis_packageIndexDepths.md) to determine how deep to search. The default depth is only 1 for most things though, so if you want your specific library to be indexed, you might need to change the packageIndexDepth. 

It also only searches for [public symbols](https://typing.python.org/en/latest/spec/distributing.html#library-interface-public-and-private-symbols)

**Reply by rchiodo** [reply]
This issue has been fixed in prerelease version 2026.1.101, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#20261101-6-March-2026-PreRelease)

**Reply by rchiodo** [reply]
Here's a picture of it working:

<img width="695" height="505" alt="image" src="https://github.com/user-attachments/assets/07ad021e-e291-48bc-99e9-e40234db2eab" />


Picking one of those jumps to the definition in the site-packages folder:

<img width="902" height="487" alt="image" src="https://github.com/user-attachments/assets/17bcb760-82e7-4025-93d0-9ded0722f14f" />
